### ISSUE TITLE
Appium XCUITest iOSDriver sendkeys() not working

### ISSUE THREAD

**Issue by tstutson** [appium/appium#12618]
TLDR: 
`sendKeys()` not working in web application text field, but only with IOSDriver/XCUITest and only on my web app's page. When running the same test in an Android simulator it works fine on my page, and everywhere else that I've tested (google and ask.com search inputs), but when I test in an iOS simulator, `sendKeys()` does not work on my web app's page, but works everywhere else (google and ask.com) and I have no idea why.

Capabilities config file:

`["Safari", "12.2", "iOS", "iPhone Simulator", "XCUITest", "1.12.1"]`

The test:
```
public class AppiumFieldsTest extends TestBase {
    @Test(dataProvider = "appium", groups = "Appium", description = "appium fields test")
    public void appiumFieldsTest(String browser, String version, String platform, String device, Method method, String automationName, String appiumVersion) throws Exception {

        IOSDriver<WebElement> driver;
        DesiredCapabilities caps = new DesiredCapabilities();
        caps.setCapability(MobileCapabilityType.PLATFORM_NAME, platform);
        caps.setCapability(MobileCapabilityType.PLATFORM_VERSION, version);
        caps.setCapability(MobileCapabilityType.DEVICE_NAME, device);
        caps.setCapability(MobileCapabilityType.AUTOMATION_NAME, automationName);
        caps.setCapability(MobileCapabilityType.BROWSER_NAME, browser);
        caps.setCapability("autoAcceptAlerts", true);
        caps.setCapability("connectHardwareKeyboard", false);
        caps.setCapability("sendKeyStrategy", "oneByOne");
        System.out.println(caps);
        String url = "http://127.0.0.1:4723/wd/hub";
        driver = new IOSDriver(new URL(url), caps);
        driver.manage().timeouts().implicitlyWait(2, TimeUnit.SECONDS);

driver.get("https://dev-app.pactsafe.com/sign?r=5cd32b47e89fc12f110449ca&s=5b2a6a597a7a3c1e4fa39c0b&signature=kto0Xohrz52ss7kc5z6t0grRbzUPsg6TrfbCbKtRuC5nQM82lNEhFL-zPgN7LaTvGG8mhuifNSc0nayvch1Rgc858Ptx8yRRD9MWJSoD4mEuHFg7LmJ-FHP~UsVEypv-gwwy-6N14BnhdkN94OZ73Kq9mBfS8QGlYKTqa76uclW0FIdnclRfA8NvK0z8CxjPcA8Luv9orw6Ye7wEuHAGqhqFURa15WeFrjrFKW9PNf6NkLVURNvOwqH4xBsfJubCkETMfjtnD4xT7PFSpgykAuU-Av0HehxCFNCYaHmyj5qvB3l9h7xgm8KKoSOO0c9VH1HpnLtwG6KAwwItawcsjg__");
        String textFieldtext = "some random text";
        Thread.sleep(10000);
        WebElement ele = driver.findElementByXPath("//*[@data-name=\"field-5b6305f656bcff936a3c53ca\"]");
        ele.click();
        Thread.sleep(3000);
        ele.sendKeys(textFieldtext);
        Assert.assertEquals(ele.getAttribute("data-value"), textFieldtext);
...
...
...
```

I am trying to do a `.click()` then a  `.sendKeys()` on the first field at the top of the page, but it is not sending any text. If you run the test you can see that the field is actually being clicked, as it turns a darker blue color, as expected, but then then the `sendKeys()` function does nothing. Also, when testing on Android, the keyboard opens on `click()` but not on iOS.

- I've tried all the different `sendKeyStratey` capabilities and still nothing. 
- I've tried different iOS simulators and appium/safari versions, and still get the same results. 
- I've tried setting the `connectHardwareKeyboard` to `true`, as well as `false`, and still the same results.
- I've added `Thread.sleep()` in-between commands - nuthin
- And again, testing on android works fine on my app, google and ask, but when testing on iOS, my app does not work, but google and ask do

The appium logs show that it is indeed trying to send the text after successfully finding the element:
```
[HTTP] {"id":"5000","text":"some random text","value":["s","o","m","e"," ","r","a","n","d","o","m"," ","t","e","x","t"]}
[W3C (bf5882ff)] Calling AppiumDriver.setValue() with args: [["s","o","m","e"," ","r","a","n","d","o","m"," ","t","e","x","t"],"5000","bf5882ff-f1a1-4ce1-bb79-02836762cb88"]
[XCUITest] Executing command 'setValue'
```

Any and all help/suggestions are greatly appreciated

**Comment by imurchie** [comment]
Please provide the server logs as a link to a [gist](https://gist.github.com).

**Comment by mykola-mokhnach** [comment]
Might be a duplicate of https://github.com/appium/appium/issues/9002

**Comment by tstutson** [comment]
@imurchie https://gist.github.com/tstutson/cbc8cfdff0a94209eb018028206f39d3

**Comment by tstutson** [comment]
@mykola-mokhnach im not entirely sure it is, as we don't use react. theres also a chance im just not understanding how the two issues could be linked regardless. could you point me to a comment that you think would be most helpful in the other issue?

**Comment by mykola-mokhnach** [comment]
I don't know what exactly might be helpful as I don't know which tools you use for development, but the idea is, that selenium atoms might not always work, because frameworks have custom event handlers and might require you to use a custom javascript in order to perform actions like text input in web views

**Comment by tstutson** [comment]
ah, yeah we're using backbone @mykola-mokhnach

**Comment by tstutson** [comment]
the text field is actually a contenteditable div, which looks like this after it attempts to send the text: 
```
<field data-element-type="field" data-id="5b6305f656bcff936a3c53ca" data-uuid="54ebaf63-41e5-4d4c-ae43-fe023548ea77" data-name="field-5b6305f656bcff936a3c53ca" data-type="text" contenteditable="true" data-required="true" data-initialized="true" class="assigned disable-enter active" style="min-width: 219px;" tabindex="1" data-editable="true"></field>
```

it should have a `data-vaue="text"` attribute after its filled out. not sure if it is a state thing, as the text isn't even showing up in the ui either

**Comment by mykola-mokhnach** [comment]
Try to provide a custom javascript snippet to edit the field content. I don't think we can do much there from Appum side.

**Comment by tstutson** [comment]
oddly enough, im able to use `js.executeScript("arguments[0].setAttribute('data-value', 'text')", element);` successfully, but it's still not being reflected in the ui for some reason

i also tried `js.executeScript("arguments[0].value='text';", element);` and it did nothing

the appium logs are showing both commands attempting to do what theyre supposed to do:

```
[W3C (99b5fc08)] Calling AppiumDriver.execute() with args: ["arguments[0].val='some random text';",[{"ELEMENT":"5000","element-6066-11e4-a52e-4f735466cecf":"5000"}],"99b5fc08-da79-4b3e-9894-03999c7615e0"]
```
and
```
[HTTP] {"script":"arguments[0].setAttribute('data-value', 'some random text')","args":[{"ELEMENT":"5000","element-6066-11e4-a52e-4f735466cecf":"5000"}]}
```

**Comment by djangofan** [comment]
For me, I always have to toggle the 'enable keyboard' settings on my Sim  off then on again for it to start working.

**Comment by mykola-mokhnach** [comment]
> oddly enough, im able to use js.executeScript("arguments[0].setAttribute('data-value', 'text')", element); successfully, but it's still not being reflected in the ui for some reason

This is normal. The upstream framework expects some sequence of events to happen there, so your task is to properly recreate that sequence and replay it in you custom script.

Closed as not an issue.

**Comment by archanapura** [comment]
Hi, I facing issue with safari browser on ios simulator. Values entered into text field is erased immediately. Used below command 
driver.findElement(By.xpath("xpath")).sendKeys(answer);
Please help with the same!!

**Comment by saymowan** [comment]
News?

**Comment by shibupanda** [comment]
**nativeWebTap = true** capability may help.
It happened with me with Cordova App (React Component).

**Comment by lanevskiy** [comment]
guys, I've faced the same issue.
on real device my keyboard language is set to Russian.
Then once I try to sendKeys('test'). It sends, but it doesn't input anything.
but, if I change manually (by hand) the keyboard language to English.
sendKeys('test') does work properly.
Is there any workaround on how to change the keyboard layout, or how to input English symbols on "Not English" keyboard?

**Comment by saymowan** [comment]
Setting the main language in capabilities can solve it ?

https://discuss.appium.io/t/how-to-changing-app-language/10726

Em sex, 22 de mai de 2020 09:57, lanevskiy <notifications@github.com>
escreveu:

**Comment by Phoenix124** [comment]
@lanevskiy I facenate the same issue for iOS 13.5 
On the iOS 12.4 work correctly

**Comment by lanevskiy** [comment]
> Setting the main language in capabilities can solve it ? https://discuss.appium.io/t/how-to-changing-app-language/10726 Em sex, 22 de mai de 2020 09:57, lanevskiy <notifications@github.com> escreveu:
> […](#)

I've tried both language and locale, and nothing has helped.

**Comment by vduhautois** [comment]
I had the same problem recently, sendKeys worked on chrome windows, and chome android but not on ios safari.
The issue was sending text on an element which was not exactly an input, but a span with contentenable=true.

On such element, the sendKeys method for desktop and android fills the innerText and textContent attributes. On ios device (XCUITest) it fills the value attribute and innerText/textContent remain empty.

So to make it work, I finally used the JavascriptExecutor to directly change the value of the innerText attribute, and it worked.

here is an example of attributes state after a sendKeys method on different environments as a proof:
```
desktop:
16:32:44.530 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 531 - innerText=Leroy Avery
16:32:45.569 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 540 - textContent=Leroy Avery
16:32:46.609 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 549 - innerHtml=null
16:32:47.650 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 558 - value=null

ios safari:
16:40:07.647 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 513 - innerText=
16:40:09.274 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 522 - textContent=
16:40:10.994 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 531 - innerHtml=null
16:40:12.469 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 540 - value=undefinedLeroy Avery

android chrome:
16:41:55.673 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 552 - innerText=Leroy Avery
16:41:57.070 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 561 - textContent=Leroy Avery
16:41:58.637 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 570 - innerHtml=null
16:41:59.990 [ForkJoinPool-1-worker-1] INFO  c.c.e.p.a.b.content.ContentComponent.fillInShortAnswer 579 - value=null
```
